vllm/v1/worker/gpu_model_runner.py

```diff
@@ -954,9 +954,17 @@ def execute_model(
         # Apply guided decoding bitmasks if present
         grammar_bitmask = scheduler_output.grammar_bitmask
         if grammar_bitmask is not None:
+            # NOTE: A non-blocking copy causes incorrect behaviour
```

Comment:
Can you explain why this is true? The reasoning here does not seem clear to me. Is it perhaps that we are overwriting `grammar_bitmask` with the next `fill_next_token_bitmask` before executing the kernel? If that is the case, then perhaps we could have one instance of the bitmask for each speculative token to avoid this problem.